samanthas_telegram_bot/api_clients/base/base_api_client.py

```diff
@@ -0,0 +1,274 @@
+                url=url,
+                method=method,
+                status_code=response.status_code,
+                attempts=attempts,
+                timeout=timeout,
+            )
+            await asyncio.sleep(timeout)
+
+    @staticmethod
+    async def _make_one_request(
+        method: HttpMethod,
+        url: str,
+        headers: dict[str, str] | None = None,
+        data: DataDict | None = None,
+        json_data: DataDict | None = None,
+        params: DataDict | None = None,
+    ) -> Response:
+        async with httpx.AsyncClient() as client:
+            if method == HttpMethod.GET:
+                response = await client.get(url, headers=headers, params=params)
+            elif method == HttpMethod.POST:
+                response = await client.post(
+                    url, headers=headers, params=params, data=data, json=json_data
+                )
+            else:
+                raise NotImplementedError(f"{method=} not supported")
+
+        logger.debug(
+            f"Sent {method.upper()} request to {url=} with {data=}. {response.status_code=}."
+        )
+        return response
+
+    @staticmethod
+    async def _log_retry(
+        update: Update,
+        context: CUSTOM_CONTEXT_TYPES,
+        url: str,
+        method: HttpMethod,
+        status_code: int,
+        timeout: int,
+        attempts: int,
+    ) -> None:
+        if attempts == 1:
+            await logs(
+                bot=context.bot,
+                update=update,
+                text=(
+                    f"Failed to reach {url=} with {method=} ({status_code=}). "
+                    f"Will try {MAX_ATTEMPTS_TO_GET_DATA_FROM_API - attempts} times before failing"
+                ),
+                level=LoggingLevel.WARNING,
+                needs_to_notify_admin_group=True,
+            )
+            return
+
+        if attempts == MAX_ATTEMPTS_TO_GET_DATA_FROM_API:
+            raise BaseApiClientError(
+                f"Tried to reach {url=} with {method=}. Failed "
+                f"{MAX_ATTEMPTS_TO_GET_DATA_FROM_API} times, will stop trying now."
+            )
+
+        await logs(
+            bot=context.bot,
+            update=update,
+            text=(
+                f"Failed to reach {url=} with {method=} ({status_code=}). "
+                f"Next attempt in {timeout} seconds "
+                f"({MAX_ATTEMPTS_TO_GET_DATA_FROM_API - attempts} attempts left)."
+            ),
+            level=LoggingLevel.WARNING,
+        )
+        return
+
+    @staticmethod
+    def _get_status_code_and_json(response: Response) -> tuple[int, DataDict]:
+        status_code = response.status_code
+        try:
+            response_json = response.json()
+        except AttributeError as err:
+            raise AttributeError(
+                f"Response contains no JSON. Response status code: {status_code}"
+            ) from err
+
+        logger.debug(f"JSON: {response_json}")
```

Comment:
Imagine you are making 100 requests at one moment. How you will determine in logs, to which request this `response_json` belongs to if all of them are mixed?

Reply:
Yeah, I needed to replace these calls with calls to `logs()` too. Fixed.